// @fluentui/style-utilities/lib-commonjs/utilities/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getIconClassName = exports.setIconOptions = exports.unregisterIcons = exports.registerIconAlias = exports.registerIcons = exports.getIcon = exports.buildClassMap = void 0;
var buildClassMap_1 = require("./buildClassMap");
Object.defineProperty(exports, "buildClassMap", { enumerable: true, get: function () { return buildClassMap_1.buildClassMap; } });
var icons_1 = require("./icons");
Object.defineProperty(exports, "getIcon", { enumerable: true, get: function () { return icons_1.getIcon; } });
Object.defineProperty(exports, "registerIcons", { enumerable: true, get: function () { return icons_1.registerIcons; } });
Object.defineProperty(exports, "registerIconAlias", { enumerable: true, get: function () { return icons_1.registerIconAlias; } });
Object.defineProperty(exports, "unregisterIcons", { enumerable: true, get: function () { return icons_1.unregisterIcons; } });
Object.defineProperty(exports, "setIconOptions", { enumerable: true, get: function () { return icons_1.setIconOptions; } });
var getIconClassName_1 = require("./getIconClassName");
Object.defineProperty(exports, "getIconClassName", { enumerable: true, get: function () { return getIconClassName_1.getIconClassName; } });
