// @fluentui/azure-themes/lib/AzureCustomizations.js
import { __assign, __rest } from "tslib";
import { AzureThemeDark } from './azure/AzureThemeDark';
import { AzureThemeLight } from './azure/AzureThemeLight';
import { AzureThemeHighContrastLight } from './azure/AzureThemeHighContrastLight';
import { AzureThemeHighContrastDark } from './azure/AzureThemeHighContrastDark';
var darkScopedSettings = AzureThemeDark.components, darkThemeSettings = __rest(AzureThemeDark, ["components"]);
var lightScopedSettings = AzureThemeLight.components, lightThemeSettings = __rest(AzureThemeLight, ["components"]);
var hcLightScopedSettings = AzureThemeHighContrastLight.components, hcLightThemeSettings = __rest(AzureThemeHighContrastLight, ["components"]);
var hcDarkScopedSettings = AzureThemeHighContrastDark.components, hcDarkThemeSettings = __rest(AzureThemeHighContrastDark, ["components"]);
export var AzureCustomizationsDark = {
    settings: {
        theme: darkThemeSettings,
    },
    scopedSettings: __assign({}, darkScopedSettings),
};
export var AzureCustomizationsLight = {
    settings: {
        theme: lightThemeSettings,
    },
    scopedSettings: __assign({}, lightScopedSettings),
};
export var AzureCustomizationsHighContrastLight = {
    settings: {
        theme: hcLightThemeSettings,
    },
    scopedSettings: __assign({}, hcLightScopedSettings),
};
export var AzureCustomizationsHighContrastDark = {
    settings: {
        theme: hcDarkThemeSettings,
    },
    scopedSettings: __assign({}, hcDarkScopedSettings),
};
