// @fluentui/react/lib-amd/components/Coachmark/Beak/Beak.styles.js
define(["require", "exports", "../../../Styling"], function (require, exports, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = void 0;
    function getStyles(props) {
        var _a;
        return {
            root: [
                {
                    position: 'absolute',
                    boxShadow: 'inherit',
                    border: 'none',
                    boxSizing: 'border-box',
                    transform: props.transform,
                    width: props.width,
                    height: props.height,
                    left: props.left,
                    top: props.top,
                    right: props.right,
                    bottom: props.bottom,
                },
            ],
            beak: {
                fill: props.color,
                display: 'block',
                selectors: (_a = {},
                    _a[Styling_1.HighContrastSelector] = {
                        fill: 'windowtext',
                    },
                    _a),
            },
        };
    }
    exports.getStyles = getStyles;
});
