// @fluentui/azure-themes/lib/azure/Constants.js
import { ScreenWidthMaxMedium, ScreenWidthMaxSmall, ScreenWidthMinMedium, getScreenSelector, } from '@fluentui/react/lib/Styling';
export var inputHeight = '18px';
export var buttonPadding = '0px 16px';
export var borderWidth = '1px';
export var borderRadius = '3px';
export var borderWidthError = '1px';
export var borderSolid = 'solid';
export var borderNone = 'none';
export var dropDownItemHeight = '32px';
export var fontFamily = 
// eslint-disable-next-line @fluentui/max-len
'Segoe UI, "Segoe UI Web (West European)", "Segoe UI", -apple-system, BlinkMacSystemFont, Roboto, "Helvetica Neue",sans-serif';
export var fontWeightRegular = '400';
export var fontWeightBold = '700';
export var inputControlHeight = '24px';
export var choiceGroupDimensions = '18px';
export var inputControlPadding = '12px';
export var inputControlHeightInner = '20px';
export var textAlignCenter = 'center';
export var transparent = 'transparent';
export var MinimumScreenSelector = getScreenSelector(0, ScreenWidthMaxSmall);
export var MediumScreenSelector = getScreenSelector(ScreenWidthMinMedium, ScreenWidthMaxMedium);
