// @fluentui/azure-themes/lib-commonjs/azure/styles/ColorPickerGridCell.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ColorPickerGridCellStyles = void 0;
var ColorPickerGridCellStyles = function (props) {
    var theme = props.theme, selected = props.selected, circle = props.circle;
    var semanticColors = theme.semanticColors;
    return {
        colorCell: [
            !selected && {
                selectors: {
                    ':focus': {
                        borderColor: semanticColors.controlOutline,
                        backgroundColor: semanticColors.listItemBackgroundHovered,
                    },
                    ':hover': {
                        borderColor: semanticColors.controlOutline,
                        backgroundColor: semanticColors.listItemBackgroundHovered,
                    },
                    ':active:hover': {
                        borderColor: semanticColors.controlOutline,
                        backgroundColor: semanticColors.listItemBackgroundHovered,
                    },
                },
            },
            selected && {
                borderColor: semanticColors.controlOutline,
                selectors: {
                    ':hover': [
                        !circle && {
                            outlineColor: semanticColors.controlOutline,
                        },
                        circle && {
                            borderColor: semanticColors.controlOutline,
                        },
                    ],
                },
            },
        ],
    };
};
exports.ColorPickerGridCellStyles = ColorPickerGridCellStyles;
