// @fluentui/azure-themes/lib-commonjs/azure/styles/Dialog.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DialogFooterStyles = exports.DialogContentStyles = void 0;
var AzureType_1 = require("../AzureType");
var AzureColors_1 = require("../AzureColors");
var DialogContentStyles = function (props) {
    var theme = props.theme, isLargeHeader = props.isLargeHeader;
    var semanticColors = theme.semanticColors;
    return {
        content: {
            color: semanticColors.bodyText,
            fontSize: theme.fonts.medium.fontSize,
        },
        title: [
            {
                fontSize: AzureType_1.FontSizes.size18,
            },
            isLargeHeader && {
                color: semanticColors.primaryButtonText,
            },
            !isLargeHeader && {
                color: semanticColors.bodyText,
            },
        ],
        topButton: {
            selectors: {
                '.ms-Dialog-button': {
                    color: semanticColors.bodyText,
                },
                '.ms-Dialog-button:hover': {
                    backgroundColor: AzureColors_1.BaseColors.RED_E00B1C,
                    color: AzureColors_1.BaseColors.WHITE,
                },
                '.ms-Dialog-button:active': {
                    backgroundColor: AzureColors_1.BaseColors.RED_E00B1C,
                    color: AzureColors_1.BaseColors.WHITE,
                },
            },
        },
        innerContent: {
            color: semanticColors.bodyText,
            fontSize: theme.fonts.medium.fontSize,
        },
        subText: {
            color: semanticColors.bodyText,
            fontSize: theme.fonts.medium.fontSize,
        },
    };
};
exports.DialogContentStyles = DialogContentStyles;
var DialogFooterStyles = function (props) {
    return {
        actions: {
            margin: '16px 0 0',
        },
    };
};
exports.DialogFooterStyles = DialogFooterStyles;
