// @fluentui/azure-themes/lib-commonjs/azure/styles/HoverCard.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlainCardStyles = exports.ExpandingCardStyles = void 0;
var StyleConstants = require("../Constants");
var ExpandingCardStyles = function (props) {
    var theme = props.theme;
    var semanticColors = theme.semanticColors;
    return {
        root: {
            selectors: {
                '.ms-Callout-main': {
                    borderColor: semanticColors.inputBorder,
                    borderStyle: StyleConstants.borderSolid,
                    borderWidth: StyleConstants.borderWidth,
                },
            },
        },
    };
};
exports.ExpandingCardStyles = ExpandingCardStyles;
var PlainCardStyles = function (props) {
    var theme = props.theme;
    var semanticColors = theme.semanticColors;
    return {
        root: {
            selectors: {
                '.ms-Callout-main': {
                    borderColor: semanticColors.inputBorder,
                    borderStyle: StyleConstants.borderSolid,
                    borderWidth: StyleConstants.borderWidth,
                },
            },
        },
    };
};
exports.PlainCardStyles = PlainCardStyles;
