// @fluentui/azure-themes/lib-commonjs/azure/styles/Rating.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RatingStyles = void 0;
var RatingStyles = function (props) {
    var disabled = props.disabled, readOnly = props.readOnly, theme = props.theme;
    var palette = theme.palette;
    var semanticColors = theme.semanticColors;
    return {
        root: [
            !disabled &&
                !readOnly && {
                selectors: {
                    '&:hover': {
                        selectors: {
                            '.ms-RatingStar-back': { color: palette.themePrimary },
                        },
                    },
                },
            },
        ],
        ratingStarBack: [
            {
                color: semanticColors.controlOutline,
            },
            (disabled || readOnly) && {
                color: semanticColors.controlOutlineDisabled,
            },
        ],
        ratingStarFront: {
            color: semanticColors.bodyText,
        },
        ratingButton: [
            !disabled &&
                !readOnly && {
                selectors: {
                    '&:hover ~ .ms-Rating-button': {
                        selectors: {
                            '.ms-RatingStar-back': { color: semanticColors.controlOutline },
                            '.ms-RatingStar-front': { color: semanticColors.controlOutline },
                        },
                    },
                },
            },
        ],
    };
};
exports.RatingStyles = RatingStyles;
