// @fluentui/azure-themes/lib-commonjs/azure/styles/SuggestionsItem.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuggestionItemStyles = void 0;
var SuggestionItemStyles = function (props) {
    var theme = props.theme;
    if (!theme) {
        return {};
    }
    var semanticColors = theme.semanticColors;
    return {
        root: {
            fontSize: theme.fonts.medium.fontSize,
            backgroundColor: semanticColors.bodyBackground,
            color: semanticColors.bodyText,
            selectors: {
                ':hover': {
                    backgroundColor: semanticColors.listItemBackgroundHovered,
                },
            },
        },
        itemButton: {
            border: 'none',
            paddingRight: '2px',
            color: semanticColors.bodyText,
            backgroundColor: semanticColors.bodyBackground,
            selectors: {
                ':hover': {
                    backgroundColor: semanticColors.listItemBackgroundHovered,
                    border: 'none',
                },
                ':active': {
                    border: 'none',
                    color: semanticColors.bodyText,
                    background: semanticColors.listItemBackgroundChecked,
                },
            },
        },
    };
};
exports.SuggestionItemStyles = SuggestionItemStyles;
