// @fluentui/azure-themes/lib-commonjs/azure/styles/TagItem.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TagItemStyles = void 0;
var Constants_1 = require("../Constants");
var TagItemStyles = function (props) {
    var theme = props.theme, selected = props.selected;
    if (!theme) {
        return {};
    }
    var semanticColors = theme.semanticColors;
    var extendedSemanticColors = semanticColors;
    return {
        root: {
            verticalAlign: 'middle',
            backgroundColor: semanticColors.listItemBackgroundChecked,
            selectors: {
                ':hover': {
                    backgroundColor: semanticColors.listItemBackgroundCheckedHovered,
                },
            },
        },
        text: [
            selected && {
                color: semanticColors.bodyText,
            },
        ],
        close: {
            color: semanticColors.bodyText,
            verticalAlign: 'middle',
            backgroundColor: 'transparent',
            selectors: {
                '&.is-disabled': {
                    backgroundColor: Constants_1.transparent,
                },
                ':active': {
                    backgroundColor: semanticColors.listHeaderBackgroundPressed,
                    color: extendedSemanticColors.controlOutlineHovered,
                },
                ':hover': {
                    backgroundColor: semanticColors.listItemBackgroundCheckedHovered,
                    color: semanticColors.bodyText,
                },
            },
        },
    };
};
exports.TagItemStyles = TagItemStyles;
