// @fluentui/azure-themes/lib-commonjs/azure/styles/TeachingBubble.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TeachingBubbleStyles = void 0;
var AzureDepths_1 = require("../AzureDepths");
var AzureType_1 = require("../AzureType");
var TeachingBubbleStyles = function (props) {
    var theme = props.theme;
    var semanticColors = theme.semanticColors;
    var extendedSemanticColors = semanticColors;
    return {
        bodyContent: {
            color: extendedSemanticColors.teachingBubbleText,
            selectors: {
                '.ms-TeachingBubble-subText': {
                    fontSize: AzureType_1.FontSizes.size14,
                },
            },
        },
        footer: {
            color: extendedSemanticColors.teachingBubbleText,
            selectors: {
                '.ms-Button-label': {
                    lineHeight: 22,
                },
            },
        },
        header: {
            color: extendedSemanticColors.teachingBubbleText,
        },
        headline: {
            color: extendedSemanticColors.teachingBubbleText,
        },
        content: {
            backgroundColor: extendedSemanticColors.teachingBubbleBackground,
            color: extendedSemanticColors.teachingBubbleText,
            selectors: {
                '.ms-TeachingBubble-header--small': {
                    fontSize: AzureType_1.FontSizes.size14,
                },
                '.ms-TeachingBubble-header--condensed': {
                    fontSize: AzureType_1.FontSizes.size14,
                },
            },
        },
        subText: {
            color: extendedSemanticColors.teachingBubbleText,
        },
        subComponentStyles: {
            callout: {
                root: {
                    boxShadow: AzureDepths_1.Depths.depth8,
                    selectors: {
                        '.ms-Callout-main': {
                            border: 0,
                        },
                        '.ms-Callout-beak': {
                            backgroundColor: extendedSemanticColors.teachingBubbleBackground,
                            border: 0,
                        },
                    },
                },
            },
        },
        closeButton: {
            color: extendedSemanticColors.teachingBubbleText,
            margin: 1,
            selectors: {
                '&:hover': {
                    backgroundColor: extendedSemanticColors.primaryButtonBackgroundPressed + " !important",
                },
                '&:active': {
                    backgroundColor: extendedSemanticColors.primaryButtonBackgroundPressed,
                },
            },
        },
        primaryButton: {
            // backgroundColor and border color requires !important to override primary btn color
            backgroundColor: extendedSemanticColors.teachingBubbleSecondaryBackground + " !important",
            borderColor: extendedSemanticColors.teachingBubbleSecondaryBackground + " !important",
            selectors: {
                '&:focus': {
                    backgroundColor: extendedSemanticColors.teachingBubblePrimaryButtonHover,
                    color: extendedSemanticColors.primaryButtonBackground,
                },
                '&:hover': {
                    backgroundColor: extendedSemanticColors.teachingBubblePrimaryButtonHover,
                    borderColor: extendedSemanticColors.teachingBubblePrimaryButtonHover,
                },
                span: {
                    color: extendedSemanticColors.teachingBubbleBackground,
                },
                '.ms-Button-label': {
                    fontSize: theme.fonts.medium.fontSize,
                },
            },
        },
        secondaryButton: {
            backgroundColor: extendedSemanticColors.teachingBubbleBackground,
            selectors: {
                '.ms-Button-label': {
                    fontSize: theme.fonts.medium.fontSize,
                },
                '&:hover': {
                    backgroundColor: extendedSemanticColors.primaryButtonBackgroundHovered,
                },
                '&:focus': {
                    backgroundColor: extendedSemanticColors.teachingBubbleBackground,
                },
            },
        },
    };
};
exports.TeachingBubbleStyles = TeachingBubbleStyles;
