// @fluentui/azure-themes/lib-commonjs/AzureCustomizations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureCustomizationsHighContrastDark = exports.AzureCustomizationsHighContrastLight = exports.AzureCustomizationsLight = exports.AzureCustomizationsDark = void 0;
var tslib_1 = require("tslib");
var AzureThemeDark_1 = require("./azure/AzureThemeDark");
var AzureThemeLight_1 = require("./azure/AzureThemeLight");
var AzureThemeHighContrastLight_1 = require("./azure/AzureThemeHighContrastLight");
var AzureThemeHighContrastDark_1 = require("./azure/AzureThemeHighContrastDark");
var darkScopedSettings = AzureThemeDark_1.AzureThemeDark.components, darkThemeSettings = tslib_1.__rest(AzureThemeDark_1.AzureThemeDark, ["components"]);
var lightScopedSettings = AzureThemeLight_1.AzureThemeLight.components, lightThemeSettings = tslib_1.__rest(AzureThemeLight_1.AzureThemeLight, ["components"]);
var hcLightScopedSettings = AzureThemeHighContrastLight_1.AzureThemeHighContrastLight.components, hcLightThemeSettings = tslib_1.__rest(AzureThemeHighContrastLight_1.AzureThemeHighContrastLight, ["components"]);
var hcDarkScopedSettings = AzureThemeHighContrastDark_1.AzureThemeHighContrastDark.components, hcDarkThemeSettings = tslib_1.__rest(AzureThemeHighContrastDark_1.AzureThemeHighContrastDark, ["components"]);
exports.AzureCustomizationsDark = {
    settings: {
        theme: darkThemeSettings,
    },
    scopedSettings: tslib_1.__assign({}, darkScopedSettings),
};
exports.AzureCustomizationsLight = {
    settings: {
        theme: lightThemeSettings,
    },
    scopedSettings: tslib_1.__assign({}, lightScopedSettings),
};
exports.AzureCustomizationsHighContrastLight = {
    settings: {
        theme: hcLightThemeSettings,
    },
    scopedSettings: tslib_1.__assign({}, hcLightScopedSettings),
};
exports.AzureCustomizationsHighContrastDark = {
    settings: {
        theme: hcDarkThemeSettings,
    },
    scopedSettings: tslib_1.__assign({}, hcDarkScopedSettings),
};
