// @fluentui/azure-themes/lib-amd/azure/AzureThemeHighContrastLight.js
define(["require", "exports", "@fluentui/react", "./AzureColors", "./AzureType", "./Constants", "./AzureStyleSettings"], function (require, exports, react_1, AzureColors_1, AzureType_1, StyleConstants, AzureStyleSettings_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.AzureThemeHighContrastLight = void 0;
    var highContrastLightExtendedSemanticColors = {
        bodyBackground: AzureColors_1.HighContrastLightSemanticColors.background,
        bodyDivider: AzureColors_1.CommonSemanticColors.dividers.lineSeparator,
        bodyText: AzureColors_1.HighContrastLightSemanticColors.text.body,
        bodyTextHovered: AzureColors_1.HighContrastLightSemanticColors.text.bodyHovered,
        buttonBackground: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.rest.background,
        buttonBackgroundChecked: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.pressed.background,
        buttonBackgroundCheckedHovered: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.hover.background,
        ButtonBackgroundDanger: AzureColors_1.HighContrastLightSemanticColors.dangerButton.rest.background,
        ButtonBackgroundDangerHovered: AzureColors_1.HighContrastLightSemanticColors.dangerButton.hover.background,
        ButtonBackgroundDangerPressed: AzureColors_1.HighContrastLightSemanticColors.dangerButton.pressed.background,
        ButtonBackgroundDangerText: AzureColors_1.HighContrastLightSemanticColors.dangerButton.rest.text,
        ButtonBackgroundDangerTextHovered: AzureColors_1.HighContrastLightSemanticColors.dangerButton.hover.text,
        buttonBackgroundDisabled: AzureColors_1.HighContrastLightSemanticColors.disabledButton.background,
        buttonBackgroundHovered: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.hover.background,
        buttonBackgroundPressed: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.pressed.background,
        ButtonBorderDanger: AzureColors_1.HighContrastLightSemanticColors.dangerButton.rest.border,
        ButtonBorderDangerHovered: AzureColors_1.HighContrastLightSemanticColors.dangerButton.hover.border,
        ButtonBorderDangerPressed: AzureColors_1.HighContrastLightSemanticColors.dangerButton.pressed.border,
        ButtonBorderDisabled: AzureColors_1.HighContrastLightSemanticColors.disabledButton.text,
        ButtonBorderFocus: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.focus.border,
        buttonText: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.rest.text,
        buttonTextChecked: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.pressed.text,
        buttonTextCheckedHovered: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.hover.border,
        buttonTextDisabled: AzureColors_1.HighContrastLightSemanticColors.disabledButton.text,
        buttonTextHovered: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.hover.color,
        buttonTextPressed: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.pressed.text,
        checkboxBackgroundChecked: AzureColors_1.HighContrastLightSemanticColors.checkBox.checked.background,
        checkboxBackgroundHovered: AzureColors_1.HighContrastLightSemanticColors.checkBox.checked.hoverBackground,
        checkBoxBorder: AzureColors_1.HighContrastLightSemanticColors.checkBox.rest.border,
        checkboxBorderChecked: AzureColors_1.HighContrastLightSemanticColors.checkBox.checked.border,
        checkboxBorderCheckedHovered: AzureColors_1.HighContrastLightSemanticColors.checkBox.checked.hoverBorder,
        checkBoxCheck: AzureColors_1.HighContrastLightSemanticColors.checkBox.rest.check,
        checkBoxCheckedFocus: AzureColors_1.HighContrastLightSemanticColors.checkBox.rest.focus,
        checkBoxCheckHover: AzureColors_1.HighContrastLightSemanticColors.checkBox.rest.hover,
        checkBoxCheckHoverTest: AzureColors_1.HighContrastLightSemanticColors.checkBox.rest.hoverText,
        checkBoxCheckedDisabledBackground: AzureColors_1.HighContrastLightSemanticColors.checkBox.disabled.background,
        checkBoxDisabled: AzureColors_1.HighContrastLightSemanticColors.checkBox.disabled.border,
        checkBoxIndeterminateBackground: AzureColors_1.HighContrastLightSemanticColors.checkBox.rest.check,
        checkBoxIndeterminateDefaultChecked: AzureColors_1.HighContrastLightSemanticColors.checkBox.checked.default,
        choiceGroupUncheckedDotHover: AzureColors_1.HighContrastLightSemanticColors.choiceGroup.circle.hover,
        choiceGroupFocusBorder: AzureColors_1.HighContrastLightSemanticColors.choiceGroup.focus,
        commandBarBorder: AzureColors_1.HighContrastLightSemanticColors.commandBar.border,
        datePickerDisabledBorder: AzureColors_1.HighContrastLightSemanticColors.datePicker.disabled.border,
        datePickerSelectionBackground: AzureColors_1.HighContrastLightSemanticColors.primaryButton.hover.background,
        datePickerSelectionText: AzureColors_1.HighContrastLightSemanticColors.datePicker.rest.text,
        disabledBackground: AzureColors_1.CommonSemanticColors.backgrounds.disabled,
        disabledBodyText: AzureColors_1.HighContrastLightSemanticColors.text.disabled,
        errorBackground: AzureColors_1.HighContrastLightSemanticColors.controlOutlines.error,
        errorText: AzureColors_1.HighContrastLightSemanticColors.text.error,
        focusBorder: AzureColors_1.HighContrastLightSemanticColors.controlOutlines.accent,
        inputBackground: AzureColors_1.HighContrastLightSemanticColors.background,
        inputBorder: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.rest.border,
        inputBorderHovered: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.hover.border,
        inputBorderPressed: AzureColors_1.HighContrastLightSemanticColors.secondaryButton.pressed.border,
        inputPlaceholderText: AzureColors_1.HighContrastLightSemanticColors.text.placeholder,
        inputText: AzureColors_1.HighContrastLightSemanticColors.text.value,
        link: AzureColors_1.HighContrastLightSemanticColors.text.hyperlink,
        linkHovered: AzureColors_1.HighContrastLightSemanticColors.text.hyperlinkHovered,
        listBackground: AzureColors_1.HighContrastLightSemanticColors.background,
        listHeaderBackgroundPressed: AzureColors_1.HighContrastLightSemanticColors.item.hover,
        listItemBackgroundChecked: AzureColors_1.HighContrastLightSemanticColors.item.select,
        listItemBackgroundCheckedHovered: AzureColors_1.HighContrastLightSemanticColors.item.select,
        listItemBackgroundHovered: AzureColors_1.HighContrastLightSemanticColors.detailsRow.hoveredBackground,
        listItemBackgroundSelected: AzureColors_1.HighContrastLightSemanticColors.item.select,
        listItemBackgroundSelectedHovered: AzureColors_1.HighContrastLightSemanticColors.item.selectHovered,
        listLinkHovered: AzureColors_1.HighContrastLightSemanticColors.detailsRow.hoveredLink,
        listLinkRowHovered: AzureColors_1.HighContrastLightSemanticColors.detailsRow.hoveredRowLink,
        listLinkRowSelected: AzureColors_1.HighContrastLightSemanticColors.detailsRow.selectedLink,
        listLinkRowSelectedHovered: AzureColors_1.HighContrastLightSemanticColors.detailsRow.selectedHoveredLink,
        listText: AzureColors_1.HighContrastLightSemanticColors.text.list,
        menuItemBackgroundHovered: AzureColors_1.HighContrastLightSemanticColors.primaryButton.hover.background,
        menuItemBackgroundPressed: AzureColors_1.HighContrastLightSemanticColors.primaryButton.hover.background,
        primaryButtonBackground: AzureColors_1.HighContrastLightSemanticColors.primaryButton.rest.background,
        primaryButtonBackgroundDisabled: AzureColors_1.HighContrastLightSemanticColors.disabledButton.background,
        primaryButtonBackgroundHovered: AzureColors_1.HighContrastLightSemanticColors.primaryButton.hover.background,
        primaryButtonBackgroundPressed: AzureColors_1.HighContrastLightSemanticColors.primaryButton.pressed.background,
        primaryButtonBorder: AzureColors_1.HighContrastLightSemanticColors.primaryButton.rest.border,
        primaryButtonBorderDisabled: AzureColors_1.HighContrastLightSemanticColors.disabledButton.text,
        primaryButtonText: AzureColors_1.HighContrastLightSemanticColors.primaryButton.rest.text,
        primaryButtonTextDisabled: AzureColors_1.HighContrastLightSemanticColors.disabledButton.text,
        primaryButtonTextFocused: AzureColors_1.HighContrastLightSemanticColors.primaryButton.focus.text,
        primaryButtonTextHovered: AzureColors_1.HighContrastLightSemanticColors.primaryButton.hover.text,
        primaryButtonTextPressed: AzureColors_1.HighContrastLightSemanticColors.primaryButton.pressed.text,
        primaryCompoundButtonBorder: AzureColors_1.HighContrastLightSemanticColors.primaryButton.rest.border,
        radioButtonPillBorderDisabled: AzureColors_1.HighContrastLightSemanticColors.radioButton.circle.borderDisabled,
        radioButtonPillCheckedHover: AzureColors_1.HighContrastLightSemanticColors.radioButton.pill.checkedHover,
        radioButtonPillUncheckedDisabled: AzureColors_1.HighContrastLightSemanticColors.radioButton.pill.uncheckedDisabled,
        radioButtonPillUncheckedHover: AzureColors_1.HighContrastLightSemanticColors.radioButton.pill.uncheckedHover,
        radioButtonPillDisabled: AzureColors_1.HighContrastLightSemanticColors.radioButton.pill.disabled,
        radioButtonThumbUnchecked: AzureColors_1.HighContrastLightSemanticColors.radioButton.circle.uncheckedRest,
        radioButtonThumbUncheckedDisabled: AzureColors_1.HighContrastLightSemanticColors.radioButton.circle.checkedDisabled,
        radioButtonThumbCheckedDisabled: AzureColors_1.HighContrastLightSemanticColors.radioButton.circle.checkedDisabled,
        rowBorder: AzureColors_1.HighContrastLightSemanticColors.radioButton.circle.checkedDisabled,
        rowFocus: AzureColors_1.HighContrastLightSemanticColors.detailsRow.focus,
        tabHover: AzureColors_1.HighContrastLightSemanticColors.tabs.hover,
        variantBorder: AzureColors_1.HighContrastLightSemanticColors.controlOutlines.rest,
        // extended
        commandBarButtonAfterColor: AzureColors_1.HighContrastLightSemanticColors.commandBar.button.focus.borderColor,
        commandBarButtonBackgroundHover: AzureColors_1.HighContrastLightSemanticColors.commandBar.button.hover.background,
        commandBarButtonBackgroundSelected: AzureColors_1.HighContrastLightSemanticColors.commandBar.button.selected.background,
        commandBarButtonBackgroundSelectedHover: AzureColors_1.HighContrastLightSemanticColors.commandBar.button.selectedHover.background,
        commandBarButtonIconHover: AzureColors_1.HighContrastLightSemanticColors.commandBar.button.hover.icon,
        commandBarButtonIconSelected: AzureColors_1.HighContrastLightSemanticColors.commandBar.button.selected.icon,
        commandBarButtonText: AzureColors_1.HighContrastLightSemanticColors.commandBar.button.root.color,
        commandBarButtonTextDisabled: AzureColors_1.HighContrastLightSemanticColors.commandBar.button.disabled.color,
        commandBarButtonTextHover: AzureColors_1.HighContrastLightSemanticColors.commandBar.button.hover.color,
        controlAccent: AzureColors_1.HighContrastLightSemanticColors.controlOutlines.accent,
        controlBackground: AzureColors_1.HighContrastLightSemanticColors.controlOutlines.background,
        controlFocus: AzureColors_1.HighContrastLightSemanticColors.controlOutlines.focus,
        controlOutline: AzureColors_1.HighContrastLightSemanticColors.controlOutlines.rest,
        controlOutlineDisabled: AzureColors_1.HighContrastLightSemanticColors.controlOutlines.disabled,
        controlOutlineHovered: AzureColors_1.HighContrastLightSemanticColors.controlOutlines.hover,
        iconButtonFill: AzureColors_1.HighContrastLightSemanticColors.text.icon,
        iconButtonFillHovered: AzureColors_1.HighContrastLightSemanticColors.primaryButton.hover.text,
        labelText: AzureColors_1.HighContrastLightSemanticColors.text.label,
        sliderActiveBackground: AzureColors_1.HighContrastLightSemanticColors.slider.activeBackground,
        sliderInActiveHover: AzureColors_1.HighContrastLightSemanticColors.slider.inactiveBackgroundHovered,
        sliderActiveHover: AzureColors_1.HighContrastLightSemanticColors.slider.activeBackgroundHovered,
        sliderActivePressed: AzureColors_1.HighContrastLightSemanticColors.slider.activeBackgroundPressed,
        sliderDisabledActiveBackground: AzureColors_1.HighContrastLightSemanticColors.slider.activeDisabledBackground,
        sliderDisabledInActiveBackground: AzureColors_1.HighContrastLightSemanticColors.slider.inactiveDisabledBackground,
        statusDefaultBackground: AzureColors_1.HighContrastLightSemanticColors.statusBar.background.default,
        statusDefaultBorder: AzureColors_1.HighContrastLightSemanticColors.statusBar.border.default,
        statusErrorBackground: AzureColors_1.HighContrastLightSemanticColors.statusBar.background.error,
        statusErrorBorder: AzureColors_1.HighContrastLightSemanticColors.statusBar.border.error,
        statusErrorText: AzureColors_1.HighContrastLightSemanticColors.text.body,
        statusErrorIcon: AzureColors_1.HighContrastLightSemanticColors.statusBar.icon.error,
        statusInformationBackground: AzureColors_1.HighContrastLightSemanticColors.statusBar.background.information,
        statusInformationText: AzureColors_1.HighContrastLightSemanticColors.text.body,
        statusInformationIcon: AzureColors_1.HighContrastLightSemanticColors.statusBar.icon.default,
        statusLink: AzureColors_1.HighContrastLightSemanticColors.statusBar.link,
        statusSuccessBackground: AzureColors_1.HighContrastLightSemanticColors.statusBar.background.okay,
        statusSuccessBorder: AzureColors_1.HighContrastLightSemanticColors.statusBar.border.okay,
        statusSuccessText: AzureColors_1.HighContrastLightSemanticColors.text.body,
        statusSuccessIcon: AzureColors_1.HighContrastLightSemanticColors.statusBar.icon.okay,
        statusWarningBackground: AzureColors_1.HighContrastLightSemanticColors.statusBar.background.warning,
        statusWarningBorder: AzureColors_1.HighContrastLightSemanticColors.statusBar.border.warning,
        statusWarningText: AzureColors_1.HighContrastLightSemanticColors.text.body,
        statusWarningIcon: AzureColors_1.HighContrastLightSemanticColors.statusBar.icon.warning,
        teachingBubbleBackground: AzureColors_1.HighContrastLightSemanticColors.teachingBubble.rest.background,
        teachingBubbleBorder: AzureColors_1.HighContrastLightSemanticColors.teachingBubble.rest.border,
        teachingBubblePrimaryButtonHover: AzureColors_1.HighContrastLightSemanticColors.teachingBubble.hover.primaryButtonBackground,
        teachingBubbleSecondaryBackground: AzureColors_1.HighContrastLightSemanticColors.teachingBubble.rest.secondaryBackround,
        teachingBubbleText: AzureColors_1.HighContrastLightSemanticColors.teachingBubble.rest.text,
        textFieldBorderDisabled: AzureColors_1.HighContrastLightSemanticColors.text.disabled,
        calendarBackground: AzureColors_1.HighContrastLightSemanticColors.calendar.background,
        calendarTextRest: AzureColors_1.HighContrastLightSemanticColors.calendar.text.rest,
        calendarTextHover: AzureColors_1.HighContrastLightSemanticColors.calendar.text.hover,
        calendarTextDisabled: AzureColors_1.HighContrastLightSemanticColors.calendar.text.disabled,
        calendarTextOutside: AzureColors_1.HighContrastLightSemanticColors.calendar.text.outside,
        calendarBackgroundBorder: AzureColors_1.HighContrastLightSemanticColors.calendar.button.border,
        calendarBackgroundSelected: AzureColors_1.HighContrastLightSemanticColors.calendar.button.selected,
        calendarBackgroundHover: AzureColors_1.HighContrastLightSemanticColors.calendar.button.hover,
        // temporary work around for high contrast themes
        choiceGroupContainerBorder: '1px',
        callOutBorderStyle: 'solid',
        choiceGroupContainerBorderStyle: 'dashed',
        listUnderline: 'underline',
        linkBorderStyle: 'dashed',
    };
    exports.AzureThemeHighContrastLight = react_1.createTheme({
        fonts: {
            medium: {
                fontFamily: StyleConstants.fontFamily,
                fontSize: AzureType_1.FontSizes.size13,
            },
            large: {
                fontSize: AzureType_1.FontSizes.size14,
            },
        },
        palette: {
            themePrimary: AzureColors_1.HighContrastLightSemanticColors.controlOutlines.accent,
            neutralPrimary: AzureColors_1.HighContrastLightSemanticColors.text.body,
            neutralDark: AzureColors_1.HighContrastLightSemanticColors.text.body,
            neutralLight: AzureColors_1.HighContrastLightSemanticColors.shimmer.secondary,
            neutralLighter: AzureColors_1.HighContrastLightSemanticColors.shimmer.primary,
            neutralLighterAlt: AzureColors_1.HighContrastLightSemanticColors.item.hover,
            neutralQuaternaryAlt: AzureColors_1.HighContrastLightSemanticColors.item.select,
            neutralSecondary: AzureColors_1.HighContrastLightSemanticColors.text.label,
            white: AzureColors_1.HighContrastLightSemanticColors.background, // shimmer elements
        },
        semanticColors: highContrastLightExtendedSemanticColors,
    });
    exports.AzureThemeHighContrastLight.components = AzureStyleSettings_1.AzureStyleSettings(exports.AzureThemeHighContrastLight);
});
