// @fluentui/azure-themes/lib-amd/azure/AzureStyleSettings.js
define(["require", "exports", "./styles/ActionButton.styles", "./styles/Breadcrumb.styles", "./styles/Callout.styles", "./styles/Checkbox.styles", "./styles/DetailsList.styles", "./styles/ChoiceGroupOptions.styles", "./styles/ColorPickerGridCell.styles", "./styles/ColorPicker.styles", "./styles/ComboBox.styles", "./styles/CommandBar.styles", "./styles/CommandBarButton.styles", "./styles/CompoundButton.styles", "./styles/ContextualMenu.styles", "./styles/DatePicker.styles", "./styles/DefaultButton.styles", "./styles/Dialog.styles", "./styles/DocumentCard.styles", "./styles/DropDown.styles", "./styles/IconButton.styles", "./styles/Label.styles", "./styles/Link.styles", "./styles/MessageBar.styles", "./styles/Modal.styles", "./styles/Overlay.styles", "./styles/Nav.styles", "./styles/Panel.styles", "./styles/Pivot.styles", "./styles/HoverCard.styles", "./styles/PrimaryButton.styles", "./styles/ProgressIndicator.styles", "./styles/Rating.styles", "./styles/SearchBox.styles", "./styles/Slider.styles", "./styles/SpinButton.styles", "./styles/SuggestionsItem.styles", "./styles/Suggestions.styles", "./styles/TagItem.styles", "./styles/TagPicker.styles", "./styles/TeachingBubble.styles", "./styles/TextField.styles", "./styles/Toggle.styles", "./styles/Tooltip.styles", "./styles/Calendar.styles"], function (require, exports, ActionButton_styles_1, Breadcrumb_styles_1, Callout_styles_1, Checkbox_styles_1, DetailsList_styles_1, ChoiceGroupOptions_styles_1, ColorPickerGridCell_styles_1, ColorPicker_styles_1, ComboBox_styles_1, CommandBar_styles_1, CommandBarButton_styles_1, CompoundButton_styles_1, ContextualMenu_styles_1, DatePicker_styles_1, DefaultButton_styles_1, Dialog_styles_1, DocumentCard_styles_1, DropDown_styles_1, IconButton_styles_1, Label_styles_1, Link_styles_1, MessageBar_styles_1, Modal_styles_1, Overlay_styles_1, Nav_styles_1, Panel_styles_1, Pivot_styles_1, HoverCard_styles_1, PrimaryButton_styles_1, ProgressIndicator_styles_1, Rating_styles_1, SearchBox_styles_1, Slider_styles_1, SpinButton_styles_1, SuggestionsItem_styles_1, Suggestions_styles_1, TagItem_styles_1, TagPicker_styles_1, TeachingBubble_styles_1, TextField_styles_1, Toggle_styles_1, Tooltip_styles_1, Calendar_styles_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.AzureStyleSettings = void 0;
    // TODO: "any" is used here to get around "is using xxx but cannot be named" TS error. Should be able to remove
    //        this 'any' once we upgrade to TS3.1+
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    var AzureStyleSettings = function (theme) {
        return {
            ActionButton: {
                styles: ActionButton_styles_1.ActionButtonStyles(theme),
            },
            Breadcrumb: {
                styles: Breadcrumb_styles_1.BreadcrumbStyles,
            },
            Calendar: {
                styles: Calendar_styles_1.CalendarStyles,
            },
            CalendarDayGrid: {
                styles: Calendar_styles_1.CalendarDayGridStyles,
            },
            CalendarDay: {
                styles: Calendar_styles_1.CalendarDayStyles,
            },
            CalendarMonth: {
                styles: Calendar_styles_1.CalendarMonthStyles,
            },
            CalendarYear: {
                styles: Calendar_styles_1.CalendarYearStyles,
            },
            CalloutContent: {
                styles: Callout_styles_1.CalloutContentStyles,
            },
            ColorPicker: {
                styles: ColorPicker_styles_1.ColorPickerStyles,
            },
            ColorRectangle: {
                styles: ColorPicker_styles_1.ColorRectangleStyles,
            },
            ColorSlider: {
                styles: ColorPicker_styles_1.ColorSliderStyles,
            },
            CommandBar: {
                styles: CommandBar_styles_1.CommandBarStyles,
            },
            CommandBarButton: {
                styles: CommandBarButton_styles_1.CommandBarButtonStyles(theme),
            },
            CompoundButton: {
                styles: CompoundButton_styles_1.CompoundButtonStyles(theme),
            },
            Check: {
                styles: DetailsList_styles_1.CheckStyles,
            },
            Checkbox: {
                styles: Checkbox_styles_1.CheckboxStyles,
            },
            ChoiceGroupOption: {
                styles: ChoiceGroupOptions_styles_1.ChoiceGroupOptionStyles,
            },
            ColorPickerGridCell: {
                styles: ColorPickerGridCell_styles_1.ColorPickerGridCellStyles,
            },
            ComboBox: {
                styles: ComboBox_styles_1.ComboBoxStyles(theme),
            },
            ContextualMenu: {
                styles: ContextualMenu_styles_1.ContextualMenuStyles,
            },
            DatePicker: {
                styles: DatePicker_styles_1.DatePickerStyles,
            },
            DefaultButton: {
                styles: DefaultButton_styles_1.DefaultButtonStyles(theme),
            },
            DetailsList: {
                styles: DetailsList_styles_1.DetailsListStyles,
            },
            DetailsRow: {
                styles: DetailsList_styles_1.DetailsRowStyles,
            },
            DialogContent: {
                styles: Dialog_styles_1.DialogContentStyles,
            },
            DialogFooter: {
                styles: Dialog_styles_1.DialogFooterStyles,
            },
            DocumentCard: {
                styles: DocumentCard_styles_1.DocumentCardStyles,
            },
            Dropdown: {
                styles: DropDown_styles_1.DropdownStyles,
            },
            ExpandingCard: {
                styles: HoverCard_styles_1.ExpandingCardStyles,
            },
            IconButton: {
                styles: IconButton_styles_1.IconButtonStyles(theme),
            },
            Label: {
                styles: Label_styles_1.LabelStyles,
            },
            Link: {
                styles: Link_styles_1.LinkStyles,
            },
            MessageBar: {
                styles: MessageBar_styles_1.MessageBarStyles,
            },
            Modal: {
                styles: Modal_styles_1.ModalStyles,
            },
            Overlay: {
                styles: Overlay_styles_1.OverlayStyles,
            },
            Nav: {
                styles: Nav_styles_1.NavStyles,
            },
            Pivot: {
                styles: Pivot_styles_1.PivotStyles,
            },
            PlainCard: {
                styles: HoverCard_styles_1.PlainCardStyles,
            },
            Panel: {
                styles: Panel_styles_1.PanelStyles,
            },
            PrimaryButton: {
                styles: PrimaryButton_styles_1.PrimaryButtonStyles(theme),
            },
            ProgressIndicator: {
                styles: ProgressIndicator_styles_1.ProgressIndicatorStyles,
            },
            Rating: {
                styles: Rating_styles_1.RatingStyles,
            },
            Slider: {
                styles: Slider_styles_1.SliderStyles,
            },
            SearchBox: {
                styles: SearchBox_styles_1.SearchBoxStyles,
            },
            SpinButton: {
                styles: SpinButton_styles_1.SpinButtonStyles,
            },
            TagPicker: {
                styles: TagPicker_styles_1.TagPickerStyles,
            },
            TagItem: {
                styles: TagItem_styles_1.TagItemStyles,
            },
            SuggestionItem: {
                styles: SuggestionsItem_styles_1.SuggestionItemStyles,
            },
            Suggestions: {
                styles: Suggestions_styles_1.SuggestionsStyles,
            },
            TeachingBubble: {
                styles: TeachingBubble_styles_1.TeachingBubbleStyles,
            },
            TextField: {
                styles: TextField_styles_1.TextFieldStyles,
            },
            Toggle: {
                styles: Toggle_styles_1.ToggleStyles,
            },
            Tooltip: {
                styles: Tooltip_styles_1.TooltipStyles,
            },
        };
    };
    exports.AzureStyleSettings = AzureStyleSettings;
});
