// @fluentui/azure-themes/lib-amd/azure/styles/Callout.styles.js
define(["require", "exports", "../AzureDepths"], function (require, exports, AzureDepths_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.CalloutContentStyles = void 0;
    var CalloutContentStyles = function (props) {
        var theme = props.theme;
        var semanticColors = theme.semanticColors;
        var extendedSemanticColors = semanticColors;
        return {
            root: {
                boxShadow: AzureDepths_1.Depths.depth8,
            },
            calloutMain: {
                color: semanticColors.bodyText,
                fontSize: theme.fonts.medium.fontSize,
                backgroundColor: extendedSemanticColors.controlBackground,
                border: extendedSemanticColors.choiceGroupContainerBorder + "\n      " + extendedSemanticColors.callOutBorderStyle + "\n      " + extendedSemanticColors.primaryButtonBorder,
            },
            beak: {
                backgroundColor: extendedSemanticColors.controlBackground,
                border: extendedSemanticColors.choiceGroupContainerBorder + "\n      " + extendedSemanticColors.callOutBorderStyle + "\n      " + extendedSemanticColors.primaryButtonBorder,
            },
        };
    };
    exports.CalloutContentStyles = CalloutContentStyles;
});
