// @fluentui/azure-themes/lib-amd/azure/styles/ContextualMenu.styles.js
define(["require", "exports", "../AzureDepths", "../Constants"], function (require, exports, AzureDepths_1, StyleConstants) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.ContextualMenuStyles = void 0;
    var ContextualMenuStyles = function (props) {
        var theme = props.theme;
        var semanticColors = theme.semanticColors;
        var extendedSemanticColors = semanticColors;
        return {
            subComponentStyles: {
                callout: {
                    root: {
                        backgroundColor: semanticColors.inputBackground,
                        borderColor: semanticColors.inputBorder,
                        borderStyle: StyleConstants.borderSolid,
                        borderWidth: StyleConstants.borderWidth,
                        boxShadow: AzureDepths_1.Depths.depth8,
                        selectors: {
                            '.ms-Callout-beak': {
                                backgroundColor: semanticColors.inputBackground,
                            },
                        },
                    },
                },
                menuItem: function (itemStyleProps) {
                    return {
                        root: [
                            {
                                fontSize: theme.fonts.medium.fontSize,
                                selectors: {
                                    '&:hover': {
                                        backgroundColor: semanticColors.buttonBackgroundHovered,
                                        color: semanticColors.buttonTextHovered,
                                        selectors: {
                                            '.ms-ContextualMenu-icon': {
                                                color: extendedSemanticColors.iconButtonFillHovered,
                                            },
                                        },
                                    },
                                    '.ms-ContextualMenu-icon': {
                                        color: extendedSemanticColors.iconButtonFill,
                                    },
                                },
                            },
                        ],
                        splitPrimary: {
                            fontSize: theme.fonts.medium.fontSize,
                        },
                        divider: {
                            backgroundColor: semanticColors.inputBorder,
                        },
                        iconColor: {
                            color: semanticColors.focusBorder,
                        },
                        item: {
                            selectors: {
                                '.is-expanded': {
                                    backgroundColor: semanticColors.menuItemBackgroundHovered,
                                    color: semanticColors.inputText,
                                },
                            },
                        },
                    };
                },
            },
        };
    };
    exports.ContextualMenuStyles = ContextualMenuStyles;
});
