// @fluentui/azure-themes/lib-amd/azure/styles/DatePicker.styles.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.DatePickerStyles = void 0;
    var DatePickerStyles = function (props) {
        var disabled = props.disabled, theme = props.theme;
        if (!theme) {
            return {};
        }
        var semanticColors = theme.semanticColors;
        var extendedSemanticColors = semanticColors;
        return {
            icon: [
                {
                    color: extendedSemanticColors.calendarTextOutside,
                    bottom: '0px',
                    top: '0px',
                    height: '19px',
                    padding: '2px 2px 0 0',
                },
                disabled && {
                    color: extendedSemanticColors.calendarTextDisabled,
                },
            ],
            root: [
                {
                    fontSize: theme.fonts.medium.fontSize,
                    selectors: {
                        '.ms-TextField-field': {
                            lineHeight: 22,
                        },
                    },
                },
                disabled && {
                    border: 'none',
                    color: semanticColors.disabledBodyText,
                    selectors: {
                        '.ms-TextField-fieldGroup': {
                            borderColor: semanticColors.datePickerDisabledBorder,
                            borderRadius: 2,
                        },
                        '.ms-TextField-field': {
                            lineHeight: 22,
                        },
                    },
                },
            ],
        };
    };
    exports.DatePickerStyles = DatePickerStyles;
});
