// @fluentui/azure-themes/lib-amd/azure/styles/Panel.styles.js
define(["require", "exports", "../AzureColors"], function (require, exports, AzureColors_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.PanelStyles = void 0;
    var PanelStyles = function (props) {
        var theme = props.theme;
        var semanticColors = theme.semanticColors;
        return {
            main: {
                backgroundColor: semanticColors.bodyBackground,
                selectors: {
                    '&.ms-Panel-main': {
                        border: 'none',
                    },
                },
            },
            closeButton: {
                color: semanticColors.bodyText,
                selectors: {
                    '&:hover': {
                        backgroundColor: AzureColors_1.BaseColors.RED_E00B1C,
                        color: AzureColors_1.BaseColors.WHITE,
                    },
                    '&:active': {
                        backgroundColor: AzureColors_1.BaseColors.RED_E00B1C,
                        color: AzureColors_1.BaseColors.WHITE,
                    },
                },
            },
            content: {
                color: semanticColors.bodyText,
            },
            headerText: {
                color: semanticColors.bodyText,
            },
        };
    };
    exports.PanelStyles = PanelStyles;
});
