// @fluentui/azure-themes/lib-amd/azure/styles/SearchBox.styles.js
define(["require", "exports", "../Constants"], function (require, exports, StyleConstants) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.SearchBoxStyles = void 0;
    var SearchBoxStyles = function (props) {
        var theme = props.theme, hasFocus = props.hasFocus;
        var semanticColors = theme.semanticColors;
        var extendedSemanticColors = semanticColors;
        return {
            root: [
                {
                    height: StyleConstants.inputControlHeight,
                    selectors: {
                        '::after': {
                            borderColor: semanticColors.primaryButtonBorder,
                        },
                    },
                },
                hasFocus && {
                    borderColor: semanticColors.focusBorder,
                },
                !hasFocus && {
                    borderColor: semanticColors.inputBorder,
                    selectors: {
                        '&:hover': {
                            borderColor: semanticColors.inputBorderHovered,
                        },
                    },
                },
            ],
            clearButton: {
                selectors: {
                    '.ms-Button': {
                        color: semanticColors.inputText,
                        selectors: {
                            ':hover': {
                                color: semanticColors.inputText,
                            },
                            ':active': {
                                color: semanticColors.inputText,
                            },
                        },
                    },
                },
            },
            field: {
                color: semanticColors.inputText,
            },
            icon: {
                color: extendedSemanticColors.iconButtonFill,
                width: '20px',
                fontSize: theme.fonts.medium.fontSize,
            },
        };
    };
    exports.SearchBoxStyles = SearchBoxStyles;
});
