// @fluentui/azure-themes/lib-amd/AzureCustomizations.js
define(["require", "exports", "tslib", "./azure/AzureThemeDark", "./azure/AzureThemeLight", "./azure/AzureThemeHighContrastLight", "./azure/AzureThemeHighContrastDark"], function (require, exports, tslib_1, AzureThemeDark_1, AzureThemeLight_1, AzureThemeHighContrastLight_1, AzureThemeHighContrastDark_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.AzureCustomizationsHighContrastDark = exports.AzureCustomizationsHighContrastLight = exports.AzureCustomizationsLight = exports.AzureCustomizationsDark = void 0;
    var darkScopedSettings = AzureThemeDark_1.AzureThemeDark.components, darkThemeSettings = tslib_1.__rest(AzureThemeDark_1.AzureThemeDark, ["components"]);
    var lightScopedSettings = AzureThemeLight_1.AzureThemeLight.components, lightThemeSettings = tslib_1.__rest(AzureThemeLight_1.AzureThemeLight, ["components"]);
    var hcLightScopedSettings = AzureThemeHighContrastLight_1.AzureThemeHighContrastLight.components, hcLightThemeSettings = tslib_1.__rest(AzureThemeHighContrastLight_1.AzureThemeHighContrastLight, ["components"]);
    var hcDarkScopedSettings = AzureThemeHighContrastDark_1.AzureThemeHighContrastDark.components, hcDarkThemeSettings = tslib_1.__rest(AzureThemeHighContrastDark_1.AzureThemeHighContrastDark, ["components"]);
    exports.AzureCustomizationsDark = {
        settings: {
            theme: darkThemeSettings,
        },
        scopedSettings: tslib_1.__assign({}, darkScopedSettings),
    };
    exports.AzureCustomizationsLight = {
        settings: {
            theme: lightThemeSettings,
        },
        scopedSettings: tslib_1.__assign({}, lightScopedSettings),
    };
    exports.AzureCustomizationsHighContrastLight = {
        settings: {
            theme: hcLightThemeSettings,
        },
        scopedSettings: tslib_1.__assign({}, hcLightScopedSettings),
    };
    exports.AzureCustomizationsHighContrastDark = {
        settings: {
            theme: hcDarkThemeSettings,
        },
        scopedSettings: tslib_1.__assign({}, hcDarkScopedSettings),
    };
});
