// @fluentui/azure-themes/lib-commonjs/azure/AzureThemeHighContrastDark.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureThemeHighContrastDark = void 0;
var react_1 = require("@fluentui/react");
var AzureColors_1 = require("./AzureColors");
var AzureType_1 = require("./AzureType");
var StyleConstants = require("./Constants");
var AzureStyleSettings_1 = require("./AzureStyleSettings");
var highContrastDarkExtendedSemanticColors = {
    bodyBackground: AzureColors_1.HighContrastDarkSemanticColors.background,
    bodyDivider: AzureColors_1.CommonSemanticColors.dividers.lineSeparator,
    bodyText: AzureColors_1.HighContrastDarkSemanticColors.text.body,
    bodyTextHovered: AzureColors_1.HighContrastDarkSemanticColors.text.bodyHovered,
    buttonBackground: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.rest.background,
    buttonBackgroundChecked: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.pressed.background,
    buttonBackgroundCheckedHovered: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.hover.background,
    ButtonBackgroundDanger: AzureColors_1.HighContrastDarkSemanticColors.dangerButton.rest.background,
    ButtonBackgroundDangerHovered: AzureColors_1.HighContrastDarkSemanticColors.dangerButton.hover.background,
    ButtonBackgroundDangerPressed: AzureColors_1.HighContrastDarkSemanticColors.dangerButton.pressed.background,
    ButtonBackgroundDangerText: AzureColors_1.HighContrastDarkSemanticColors.dangerButton.rest.text,
    ButtonBackgroundDangerTextHovered: AzureColors_1.HighContrastDarkSemanticColors.dangerButton.hover.text,
    buttonBackgroundDisabled: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.disabled.background,
    buttonBackgroundHovered: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.hover.background,
    buttonBackgroundPressed: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.pressed.background,
    ButtonBorderDanger: AzureColors_1.HighContrastDarkSemanticColors.dangerButton.rest.border,
    ButtonBorderDangerHovered: AzureColors_1.HighContrastDarkSemanticColors.dangerButton.hover.border,
    ButtonBorderDangerPressed: AzureColors_1.HighContrastDarkSemanticColors.dangerButton.pressed.border,
    ButtonBorderDisabled: AzureColors_1.HighContrastDarkSemanticColors.disabledButton.background,
    ButtonBorderFocus: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.focus.border,
    buttonText: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.rest.text,
    buttonTextChecked: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.pressed.text,
    buttonTextCheckedHovered: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.hover.border,
    buttonTextDisabled: AzureColors_1.HighContrastDarkSemanticColors.disabledButton.text,
    buttonTextHovered: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.hover.color,
    buttonTextPressed: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.pressed.text,
    checkboxBackgroundChecked: AzureColors_1.HighContrastDarkSemanticColors.checkBox.checked.background,
    checkboxBackgroundHovered: AzureColors_1.HighContrastDarkSemanticColors.checkBox.checked.hoverBackground,
    checkBoxBorder: AzureColors_1.HighContrastDarkSemanticColors.checkBox.rest.border,
    checkboxBorderChecked: AzureColors_1.HighContrastDarkSemanticColors.checkBox.checked.border,
    checkboxBorderCheckedHovered: AzureColors_1.HighContrastDarkSemanticColors.checkBox.checked.hoverBorder,
    checkBoxCheck: AzureColors_1.HighContrastDarkSemanticColors.checkBox.rest.check,
    checkBoxCheckedFocus: AzureColors_1.HighContrastDarkSemanticColors.checkBox.rest.focus,
    checkBoxCheckHover: AzureColors_1.HighContrastDarkSemanticColors.checkBox.rest.hover,
    checkBoxCheckHoverTest: AzureColors_1.HighContrastDarkSemanticColors.checkBox.rest.hoverText,
    checkBoxCheckedDisabledBackground: AzureColors_1.HighContrastDarkSemanticColors.checkBox.disabled.background,
    checkBoxDisabled: AzureColors_1.HighContrastDarkSemanticColors.checkBox.disabled.border,
    checkBoxIndeterminateBackground: AzureColors_1.HighContrastDarkSemanticColors.checkBox.checked.background,
    checkBoxIndeterminateDefaultChecked: AzureColors_1.HighContrastDarkSemanticColors.checkBox.checked.default,
    choiceGroupUncheckedDotHover: AzureColors_1.HighContrastDarkSemanticColors.choiceGroup.circle.hover,
    choiceGroupFocusBorder: AzureColors_1.HighContrastDarkSemanticColors.choiceGroup.focus,
    commandBarBorder: AzureColors_1.HighContrastDarkSemanticColors.commandBar.border,
    datePickerDisabledBorder: AzureColors_1.HighContrastDarkSemanticColors.datePicker.disabled.border,
    datePickerSelectionBackground: AzureColors_1.HighContrastDarkSemanticColors.datePicker.rest.selected,
    datePickerSelectionText: AzureColors_1.HighContrastDarkSemanticColors.datePicker.rest.text,
    disabledBackground: AzureColors_1.CommonSemanticColors.backgrounds.disabled,
    disabledBodyText: AzureColors_1.HighContrastDarkSemanticColors.text.disabled,
    errorBackground: AzureColors_1.HighContrastDarkSemanticColors.controlOutlines.error,
    errorText: AzureColors_1.HighContrastDarkSemanticColors.text.error,
    focusBorder: AzureColors_1.HighContrastDarkSemanticColors.controlOutlines.accent,
    inputBackground: AzureColors_1.HighContrastDarkSemanticColors.background,
    inputBorder: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.rest.border,
    inputBorderHovered: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.hover.border,
    inputBorderPressed: AzureColors_1.HighContrastDarkSemanticColors.secondaryButton.pressed.border,
    inputPlaceholderText: AzureColors_1.HighContrastDarkSemanticColors.text.placeholder,
    inputText: AzureColors_1.HighContrastDarkSemanticColors.text.value,
    link: AzureColors_1.HighContrastDarkSemanticColors.text.hyperlink,
    linkHovered: AzureColors_1.HighContrastDarkSemanticColors.text.hyperlinkHovered,
    listBackground: AzureColors_1.HighContrastDarkSemanticColors.background,
    listHeaderBackgroundPressed: AzureColors_1.HighContrastDarkSemanticColors.item.hover,
    listItemBackgroundChecked: AzureColors_1.HighContrastDarkSemanticColors.item.select,
    listItemBackgroundCheckedHovered: AzureColors_1.HighContrastDarkSemanticColors.item.select,
    listItemBackgroundHovered: AzureColors_1.HighContrastDarkSemanticColors.detailsRow.hoveredBackground,
    listItemBackgroundSelected: AzureColors_1.HighContrastDarkSemanticColors.item.select,
    listItemBackgroundSelectedHovered: AzureColors_1.HighContrastDarkSemanticColors.item.selectHovered,
    listLinkHovered: AzureColors_1.HighContrastDarkSemanticColors.detailsRow.hoveredLink,
    listLinkRowHovered: AzureColors_1.HighContrastDarkSemanticColors.detailsRow.hoveredRowLink,
    listLinkRowSelected: AzureColors_1.HighContrastDarkSemanticColors.detailsRow.selectedLink,
    listLinkRowSelectedHovered: AzureColors_1.HighContrastDarkSemanticColors.detailsRow.selectedHoveredLink,
    listText: AzureColors_1.HighContrastDarkSemanticColors.text.list,
    menuItemBackgroundHovered: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.hover.background,
    menuItemBackgroundPressed: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.hover.background,
    primaryButtonBackground: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.rest.background,
    primaryButtonBackgroundDisabled: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.disabled.background,
    primaryButtonBackgroundHovered: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.hover.background,
    primaryButtonBackgroundPressed: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.pressed.background,
    primaryButtonBorder: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.rest.border,
    primaryButtonBorderDisabled: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.disabled.border,
    primaryButtonText: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.rest.text,
    primaryButtonTextDisabled: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.disabled.text,
    primaryButtonTextFocused: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.focus.text,
    primaryButtonTextHovered: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.hover.text,
    primaryButtonTextPressed: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.pressed.text,
    radioButtonPillBorderDisabled: AzureColors_1.HighContrastDarkSemanticColors.radioButton.circle.borderDisabled,
    radioButtonPillCheckedHover: AzureColors_1.HighContrastDarkSemanticColors.radioButton.pill.checkedHover,
    radioButtonPillUncheckedDisabled: AzureColors_1.HighContrastDarkSemanticColors.radioButton.pill.uncheckedDisabled,
    radioButtonPillUncheckedHover: AzureColors_1.HighContrastDarkSemanticColors.radioButton.pill.uncheckedHover,
    radioButtonPillDisabled: AzureColors_1.HighContrastDarkSemanticColors.radioButton.pill.disabled,
    radioButtonThumbUnchecked: AzureColors_1.HighContrastDarkSemanticColors.radioButton.circle.uncheckedRest,
    radioButtonThumbUncheckedDisabled: AzureColors_1.HighContrastDarkSemanticColors.radioButton.circle.checkedDisabled,
    radioButtonThumbCheckedDisabled: AzureColors_1.HighContrastDarkSemanticColors.radioButton.circle.checkedDisabled,
    tabHover: AzureColors_1.HighContrastDarkSemanticColors.tabs.hover,
    rowBorder: AzureColors_1.HighContrastDarkSemanticColors.detailsRow.border,
    rowFocus: AzureColors_1.HighContrastDarkSemanticColors.detailsRow.focus,
    variantBorder: AzureColors_1.CommonSemanticColors.dividers.lineSeparator,
    // extended
    commandBarButtonAfterColor: AzureColors_1.HighContrastDarkSemanticColors.commandBar.button.focus.borderColor,
    commandBarButtonBackgroundHover: AzureColors_1.HighContrastDarkSemanticColors.commandBar.button.hover.background,
    commandBarButtonBackgroundSelected: AzureColors_1.HighContrastDarkSemanticColors.commandBar.button.selected.background,
    commandBarButtonBackgroundSelectedHover: AzureColors_1.HighContrastDarkSemanticColors.commandBar.button.selectedHover.background,
    commandBarButtonIconHover: AzureColors_1.HighContrastDarkSemanticColors.commandBar.button.hover.icon,
    commandBarButtonIconSelected: AzureColors_1.HighContrastDarkSemanticColors.commandBar.button.selected.icon,
    commandBarButtonText: AzureColors_1.HighContrastDarkSemanticColors.commandBar.button.root.color,
    commandBarButtonTextDisabled: AzureColors_1.HighContrastDarkSemanticColors.commandBar.button.disabled.color,
    commandBarButtonTextHover: AzureColors_1.HighContrastDarkSemanticColors.commandBar.button.hover.color,
    controlAccent: AzureColors_1.HighContrastDarkSemanticColors.controlOutlines.accent,
    controlBackground: AzureColors_1.HighContrastDarkSemanticColors.controlOutlines.background,
    controlFocus: AzureColors_1.HighContrastDarkSemanticColors.controlOutlines.focus,
    controlOutline: AzureColors_1.HighContrastDarkSemanticColors.controlOutlines.rest,
    controlOutlineDisabled: AzureColors_1.HighContrastDarkSemanticColors.controlOutlines.disabled,
    controlOutlineHovered: AzureColors_1.HighContrastDarkSemanticColors.controlOutlines.hover,
    iconButtonFill: AzureColors_1.HighContrastDarkSemanticColors.text.icon,
    iconButtonFillHovered: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.hover.text,
    labelText: AzureColors_1.HighContrastDarkSemanticColors.text.label,
    sliderActiveBackground: AzureColors_1.HighContrastDarkSemanticColors.slider.activeBackground,
    sliderInActiveHover: AzureColors_1.HighContrastDarkSemanticColors.slider.inactiveBackgroundHovered,
    sliderActiveHover: AzureColors_1.HighContrastDarkSemanticColors.slider.activeBackgroundHovered,
    sliderActivePressed: AzureColors_1.HighContrastDarkSemanticColors.slider.activeBackgroundPressed,
    sliderDisabledActiveBackground: AzureColors_1.HighContrastDarkSemanticColors.slider.activeDisabledBackground,
    sliderDisabledInActiveBackground: AzureColors_1.HighContrastDarkSemanticColors.slider.inactiveDisabledBackground,
    statusDefaultBackground: AzureColors_1.HighContrastDarkSemanticColors.statusBar.background.default,
    statusDefaultBorder: AzureColors_1.HighContrastDarkSemanticColors.statusBar.border.default,
    statusErrorBackground: AzureColors_1.HighContrastDarkSemanticColors.statusBar.background.error,
    statusErrorBorder: AzureColors_1.HighContrastDarkSemanticColors.statusBar.border.error,
    statusErrorText: AzureColors_1.HighContrastDarkSemanticColors.text.body,
    statusErrorIcon: AzureColors_1.HighContrastDarkSemanticColors.statusBar.icon.error,
    statusInformationBackground: AzureColors_1.HighContrastDarkSemanticColors.statusBar.background.information,
    statusInformationText: AzureColors_1.HighContrastDarkSemanticColors.text.body,
    statusInformationIcon: AzureColors_1.HighContrastDarkSemanticColors.statusBar.icon.default,
    statusLink: AzureColors_1.HighContrastDarkSemanticColors.statusBar.link,
    statusSuccessBackground: AzureColors_1.HighContrastDarkSemanticColors.statusBar.background.okay,
    statusSuccessBorder: AzureColors_1.HighContrastDarkSemanticColors.statusBar.border.okay,
    statusSuccessText: AzureColors_1.HighContrastDarkSemanticColors.text.body,
    statusSuccessIcon: AzureColors_1.HighContrastDarkSemanticColors.statusBar.icon.okay,
    statusWarningBackground: AzureColors_1.HighContrastDarkSemanticColors.statusBar.background.warning,
    statusWarningBorder: AzureColors_1.HighContrastDarkSemanticColors.statusBar.border.warning,
    statusWarningText: AzureColors_1.HighContrastDarkSemanticColors.text.body,
    statusWarningIcon: AzureColors_1.HighContrastDarkSemanticColors.statusBar.icon.warning,
    teachingBubbleBackground: AzureColors_1.HighContrastDarkSemanticColors.teachingBubble.rest.background,
    teachingBubbleBorder: AzureColors_1.HighContrastDarkSemanticColors.teachingBubble.rest.border,
    teachingBubblePrimaryButtonHover: AzureColors_1.HighContrastDarkSemanticColors.teachingBubble.hover.primaryButtonBackground,
    teachingBubbleSecondaryBackground: AzureColors_1.HighContrastDarkSemanticColors.teachingBubble.rest.secondaryBackround,
    teachingBubbleText: AzureColors_1.HighContrastDarkSemanticColors.teachingBubble.rest.text,
    textFieldBorderDisabled: AzureColors_1.HighContrastDarkSemanticColors.primaryButton.disabled.border,
    calendarBackground: AzureColors_1.HighContrastDarkSemanticColors.calendar.background,
    calendarTextRest: AzureColors_1.HighContrastDarkSemanticColors.calendar.text.rest,
    calendarTextHover: AzureColors_1.HighContrastDarkSemanticColors.calendar.text.hover,
    calendarTextDisabled: AzureColors_1.HighContrastDarkSemanticColors.calendar.text.disabled,
    calendarTextOutside: AzureColors_1.HighContrastDarkSemanticColors.calendar.text.outside,
    calendarBackgroundBorder: AzureColors_1.HighContrastDarkSemanticColors.calendar.button.border,
    calendarBackgroundSelected: AzureColors_1.HighContrastDarkSemanticColors.calendar.button.selected,
    calendarBackgroundHover: AzureColors_1.HighContrastDarkSemanticColors.calendar.button.hover,
    // temporary work around for high contrast themes
    choiceGroupContainerBorder: '1px',
    callOutBorderStyle: 'solid',
    choiceGroupContainerBorderStyle: 'solid',
    listUnderline: 'underline',
    linkBorderStyle: 'dashed',
};
exports.AzureThemeHighContrastDark = react_1.createTheme({
    fonts: {
        medium: {
            fontFamily: StyleConstants.fontFamily,
            fontSize: AzureType_1.FontSizes.size13,
        },
        large: {
            fontSize: AzureType_1.FontSizes.size14,
        },
    },
    palette: {
        themePrimary: AzureColors_1.HighContrastDarkSemanticColors.controlOutlines.accent,
        neutralPrimary: AzureColors_1.HighContrastDarkSemanticColors.text.body,
        neutralDark: AzureColors_1.HighContrastDarkSemanticColors.text.body,
        neutralLighter: AzureColors_1.HighContrastDarkSemanticColors.shimmer.secondary,
        neutralLight: AzureColors_1.HighContrastDarkSemanticColors.shimmer.primary,
        neutralLighterAlt: AzureColors_1.HighContrastDarkSemanticColors.item.hover,
        neutralQuaternaryAlt: AzureColors_1.HighContrastDarkSemanticColors.item.select,
        neutralSecondary: AzureColors_1.HighContrastDarkSemanticColors.text.label,
        white: AzureColors_1.HighContrastDarkSemanticColors.background, // shimmer elements
    },
    semanticColors: highContrastDarkExtendedSemanticColors,
    isInverted: true,
});
exports.AzureThemeHighContrastDark.components = AzureStyleSettings_1.AzureStyleSettings(exports.AzureThemeHighContrastDark);
