// @fluentui/azure-themes/lib-commonjs/azure/AzureStyleSettings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureStyleSettings = void 0;
var ActionButton_styles_1 = require("./styles/ActionButton.styles");
var Breadcrumb_styles_1 = require("./styles/Breadcrumb.styles");
var Callout_styles_1 = require("./styles/Callout.styles");
var Checkbox_styles_1 = require("./styles/Checkbox.styles");
var DetailsList_styles_1 = require("./styles/DetailsList.styles");
var ChoiceGroupOptions_styles_1 = require("./styles/ChoiceGroupOptions.styles");
var ColorPickerGridCell_styles_1 = require("./styles/ColorPickerGridCell.styles");
var ColorPicker_styles_1 = require("./styles/ColorPicker.styles");
var ComboBox_styles_1 = require("./styles/ComboBox.styles");
var CommandBar_styles_1 = require("./styles/CommandBar.styles");
var CommandBarButton_styles_1 = require("./styles/CommandBarButton.styles");
var CompoundButton_styles_1 = require("./styles/CompoundButton.styles");
var ContextualMenu_styles_1 = require("./styles/ContextualMenu.styles");
var DatePicker_styles_1 = require("./styles/DatePicker.styles");
var DefaultButton_styles_1 = require("./styles/DefaultButton.styles");
var Dialog_styles_1 = require("./styles/Dialog.styles");
var DocumentCard_styles_1 = require("./styles/DocumentCard.styles");
var DropDown_styles_1 = require("./styles/DropDown.styles");
var IconButton_styles_1 = require("./styles/IconButton.styles");
var Label_styles_1 = require("./styles/Label.styles");
var Link_styles_1 = require("./styles/Link.styles");
var MessageBar_styles_1 = require("./styles/MessageBar.styles");
var Modal_styles_1 = require("./styles/Modal.styles");
var Overlay_styles_1 = require("./styles/Overlay.styles");
var Nav_styles_1 = require("./styles/Nav.styles");
var Panel_styles_1 = require("./styles/Panel.styles");
var Pivot_styles_1 = require("./styles/Pivot.styles");
var HoverCard_styles_1 = require("./styles/HoverCard.styles");
var PrimaryButton_styles_1 = require("./styles/PrimaryButton.styles");
var ProgressIndicator_styles_1 = require("./styles/ProgressIndicator.styles");
var Rating_styles_1 = require("./styles/Rating.styles");
var SearchBox_styles_1 = require("./styles/SearchBox.styles");
var Slider_styles_1 = require("./styles/Slider.styles");
var SpinButton_styles_1 = require("./styles/SpinButton.styles");
var SuggestionsItem_styles_1 = require("./styles/SuggestionsItem.styles");
var Suggestions_styles_1 = require("./styles/Suggestions.styles");
var TagItem_styles_1 = require("./styles/TagItem.styles");
var TagPicker_styles_1 = require("./styles/TagPicker.styles");
var TeachingBubble_styles_1 = require("./styles/TeachingBubble.styles");
var TextField_styles_1 = require("./styles/TextField.styles");
var Toggle_styles_1 = require("./styles/Toggle.styles");
var Tooltip_styles_1 = require("./styles/Tooltip.styles");
var Calendar_styles_1 = require("./styles/Calendar.styles");
// TODO: "any" is used here to get around "is using xxx but cannot be named" TS error. Should be able to remove
//        this 'any' once we upgrade to TS3.1+
// eslint-disable-next-line @typescript-eslint/no-explicit-any
var AzureStyleSettings = function (theme) {
    return {
        ActionButton: {
            styles: ActionButton_styles_1.ActionButtonStyles(theme),
        },
        Breadcrumb: {
            styles: Breadcrumb_styles_1.BreadcrumbStyles,
        },
        Calendar: {
            styles: Calendar_styles_1.CalendarStyles,
        },
        CalendarDayGrid: {
            styles: Calendar_styles_1.CalendarDayGridStyles,
        },
        CalendarDay: {
            styles: Calendar_styles_1.CalendarDayStyles,
        },
        CalendarMonth: {
            styles: Calendar_styles_1.CalendarMonthStyles,
        },
        CalendarYear: {
            styles: Calendar_styles_1.CalendarYearStyles,
        },
        CalloutContent: {
            styles: Callout_styles_1.CalloutContentStyles,
        },
        ColorPicker: {
            styles: ColorPicker_styles_1.ColorPickerStyles,
        },
        ColorRectangle: {
            styles: ColorPicker_styles_1.ColorRectangleStyles,
        },
        ColorSlider: {
            styles: ColorPicker_styles_1.ColorSliderStyles,
        },
        CommandBar: {
            styles: CommandBar_styles_1.CommandBarStyles,
        },
        CommandBarButton: {
            styles: CommandBarButton_styles_1.CommandBarButtonStyles(theme),
        },
        CompoundButton: {
            styles: CompoundButton_styles_1.CompoundButtonStyles(theme),
        },
        Check: {
            styles: DetailsList_styles_1.CheckStyles,
        },
        Checkbox: {
            styles: Checkbox_styles_1.CheckboxStyles,
        },
        ChoiceGroupOption: {
            styles: ChoiceGroupOptions_styles_1.ChoiceGroupOptionStyles,
        },
        ColorPickerGridCell: {
            styles: ColorPickerGridCell_styles_1.ColorPickerGridCellStyles,
        },
        ComboBox: {
            styles: ComboBox_styles_1.ComboBoxStyles(theme),
        },
        ContextualMenu: {
            styles: ContextualMenu_styles_1.ContextualMenuStyles,
        },
        DatePicker: {
            styles: DatePicker_styles_1.DatePickerStyles,
        },
        DefaultButton: {
            styles: DefaultButton_styles_1.DefaultButtonStyles(theme),
        },
        DetailsList: {
            styles: DetailsList_styles_1.DetailsListStyles,
        },
        DetailsRow: {
            styles: DetailsList_styles_1.DetailsRowStyles,
        },
        DialogContent: {
            styles: Dialog_styles_1.DialogContentStyles,
        },
        DialogFooter: {
            styles: Dialog_styles_1.DialogFooterStyles,
        },
        DocumentCard: {
            styles: DocumentCard_styles_1.DocumentCardStyles,
        },
        Dropdown: {
            styles: DropDown_styles_1.DropdownStyles,
        },
        ExpandingCard: {
            styles: HoverCard_styles_1.ExpandingCardStyles,
        },
        IconButton: {
            styles: IconButton_styles_1.IconButtonStyles(theme),
        },
        Label: {
            styles: Label_styles_1.LabelStyles,
        },
        Link: {
            styles: Link_styles_1.LinkStyles,
        },
        MessageBar: {
            styles: MessageBar_styles_1.MessageBarStyles,
        },
        Modal: {
            styles: Modal_styles_1.ModalStyles,
        },
        Overlay: {
            styles: Overlay_styles_1.OverlayStyles,
        },
        Nav: {
            styles: Nav_styles_1.NavStyles,
        },
        Pivot: {
            styles: Pivot_styles_1.PivotStyles,
        },
        PlainCard: {
            styles: HoverCard_styles_1.PlainCardStyles,
        },
        Panel: {
            styles: Panel_styles_1.PanelStyles,
        },
        PrimaryButton: {
            styles: PrimaryButton_styles_1.PrimaryButtonStyles(theme),
        },
        ProgressIndicator: {
            styles: ProgressIndicator_styles_1.ProgressIndicatorStyles,
        },
        Rating: {
            styles: Rating_styles_1.RatingStyles,
        },
        Slider: {
            styles: Slider_styles_1.SliderStyles,
        },
        SearchBox: {
            styles: SearchBox_styles_1.SearchBoxStyles,
        },
        SpinButton: {
            styles: SpinButton_styles_1.SpinButtonStyles,
        },
        TagPicker: {
            styles: TagPicker_styles_1.TagPickerStyles,
        },
        TagItem: {
            styles: TagItem_styles_1.TagItemStyles,
        },
        SuggestionItem: {
            styles: SuggestionsItem_styles_1.SuggestionItemStyles,
        },
        Suggestions: {
            styles: Suggestions_styles_1.SuggestionsStyles,
        },
        TeachingBubble: {
            styles: TeachingBubble_styles_1.TeachingBubbleStyles,
        },
        TextField: {
            styles: TextField_styles_1.TextFieldStyles,
        },
        Toggle: {
            styles: Toggle_styles_1.ToggleStyles,
        },
        Tooltip: {
            styles: Tooltip_styles_1.TooltipStyles,
        },
    };
};
exports.AzureStyleSettings = AzureStyleSettings;
