// @fluentui/azure-themes/lib-commonjs/azure/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediumScreenSelector = exports.MinimumScreenSelector = exports.transparent = exports.textAlignCenter = exports.inputControlHeightInner = exports.inputControlPadding = exports.choiceGroupDimensions = exports.inputControlHeight = exports.fontWeightBold = exports.fontWeightRegular = exports.fontFamily = exports.dropDownItemHeight = exports.borderNone = exports.borderSolid = exports.borderWidthError = exports.borderRadius = exports.borderWidth = exports.buttonPadding = exports.inputHeight = void 0;
var Styling_1 = require("@fluentui/react/lib/Styling");
exports.inputHeight = '18px';
exports.buttonPadding = '0px 16px';
exports.borderWidth = '1px';
exports.borderRadius = '3px';
exports.borderWidthError = '1px';
exports.borderSolid = 'solid';
exports.borderNone = 'none';
exports.dropDownItemHeight = '32px';
exports.fontFamily = 
// eslint-disable-next-line @fluentui/max-len
'Segoe UI, "Segoe UI Web (West European)", "Segoe UI", -apple-system, BlinkMacSystemFont, Roboto, "Helvetica Neue",sans-serif';
exports.fontWeightRegular = '400';
exports.fontWeightBold = '700';
exports.inputControlHeight = '24px';
exports.choiceGroupDimensions = '18px';
exports.inputControlPadding = '12px';
exports.inputControlHeightInner = '20px';
exports.textAlignCenter = 'center';
exports.transparent = 'transparent';
exports.MinimumScreenSelector = Styling_1.getScreenSelector(0, Styling_1.ScreenWidthMaxSmall);
exports.MediumScreenSelector = Styling_1.getScreenSelector(Styling_1.ScreenWidthMinMedium, Styling_1.ScreenWidthMaxMedium);
