// @fluentui/merge-styles/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setRTL = exports.keyframes = exports.fontFace = exports.concatStyleSetsWithProps = exports.concatStyleSets = exports.mergeCssSets = exports.mergeStyleSets = exports.mergeCss = exports.mergeStyles = void 0;
var tslib_1 = require("tslib");
var mergeStyles_1 = require("./mergeStyles");
Object.defineProperty(exports, "mergeStyles", { enumerable: true, get: function () { return mergeStyles_1.mergeStyles; } });
Object.defineProperty(exports, "mergeCss", { enumerable: true, get: function () { return mergeStyles_1.mergeCss; } });
var mergeStyleSets_1 = require("./mergeStyleSets");
Object.defineProperty(exports, "mergeStyleSets", { enumerable: true, get: function () { return mergeStyleSets_1.mergeStyleSets; } });
Object.defineProperty(exports, "mergeCssSets", { enumerable: true, get: function () { return mergeStyleSets_1.mergeCssSets; } });
var concatStyleSets_1 = require("./concatStyleSets");
Object.defineProperty(exports, "concatStyleSets", { enumerable: true, get: function () { return concatStyleSets_1.concatStyleSets; } });
var concatStyleSetsWithProps_1 = require("./concatStyleSetsWithProps");
Object.defineProperty(exports, "concatStyleSetsWithProps", { enumerable: true, get: function () { return concatStyleSetsWithProps_1.concatStyleSetsWithProps; } });
var fontFace_1 = require("./fontFace");
Object.defineProperty(exports, "fontFace", { enumerable: true, get: function () { return fontFace_1.fontFace; } });
var keyframes_1 = require("./keyframes");
Object.defineProperty(exports, "keyframes", { enumerable: true, get: function () { return keyframes_1.keyframes; } });
tslib_1.__exportStar(require("./Stylesheet"), exports);
var StyleOptionsState_1 = require("./StyleOptionsState");
Object.defineProperty(exports, "setRTL", { enumerable: true, get: function () { return StyleOptionsState_1.setRTL; } });
require("./version");
