// @fluentui/react/lib-amd/components/Shimmer/ShimmerGap/ShimmerGap.styles.js
define(["require", "exports", "../../../Styling"], function (require, exports, Styling_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.getStyles = void 0;
    var GlobalClassNames = {
        root: 'ms-ShimmerGap-root',
    };
    function getStyles(props) {
        var _a;
        // eslint-disable-next-line deprecation/deprecation
        var height = props.height, borderStyle = props.borderStyle, theme = props.theme;
        var semanticColors = theme.semanticColors;
        var globalClassNames = Styling_1.getGlobalClassNames(GlobalClassNames, theme);
        var borderStyles = borderStyle || {};
        return {
            root: [
                globalClassNames.root,
                theme.fonts.medium,
                {
                    backgroundColor: semanticColors.bodyBackground,
                    height: height + "px",
                    boxSizing: 'content-box',
                    borderTopStyle: 'solid',
                    borderBottomStyle: 'solid',
                    borderColor: semanticColors.bodyBackground,
                    selectors: (_a = {},
                        _a[Styling_1.HighContrastSelector] = {
                            backgroundColor: 'Window',
                            borderColor: 'Window',
                        },
                        _a),
                },
                borderStyles,
            ],
        };
    }
    exports.getStyles = getStyles;
});
