// @fluentui/azure-themes/lib/azure/styles/ColorPicker.styles.js
import { Depths } from '../AzureDepths';
export var ColorPickerStyles = function (props) {
    return {
        input: {
            selectors: {
                '&.ms-TextField': {
                    paddingRight: 4,
                },
                '.ms-TextField-field': {
                    minWidth: 'auto',
                    padding: 5,
                    textOverflow: 'clip',
                },
            },
        },
        table: {
            selectors: {
                'tbody td:last-of-type .ms-ColorPicker-input': {
                    paddingRight: 0,
                },
            },
        },
        tableHeader: {
            selectors: {
                td: {
                    paddingBottom: 4,
                },
            },
        },
    };
};
export var ColorRectangleStyles = function (props) {
    var theme = props.theme;
    var semanticColors = theme.semanticColors;
    return {
        root: {
            border: "1px solid " + semanticColors.controlOutline,
        },
        thumb: {
            borderColor: semanticColors.controlOutline,
            boxShadow: Depths.depth8,
        },
    };
};
export var ColorSliderStyles = function (props) {
    var theme = props.theme;
    var semanticColors = theme.semanticColors;
    return {
        sliderThumb: {
            borderColor: semanticColors.controlOutline,
            boxShadow: Depths.depth8,
        },
    };
};
