// @fluentui/azure-themes/lib/azure/styles/Dialog.styles.js
import { FontSizes } from '../AzureType';
import { BaseColors } from '../AzureColors';
export var DialogContentStyles = function (props) {
    var theme = props.theme, isLargeHeader = props.isLargeHeader;
    var semanticColors = theme.semanticColors;
    return {
        content: {
            color: semanticColors.bodyText,
            fontSize: theme.fonts.medium.fontSize,
        },
        title: [
            {
                fontSize: FontSizes.size18,
            },
            isLargeHeader && {
                color: semanticColors.primaryButtonText,
            },
            !isLargeHeader && {
                color: semanticColors.bodyText,
            },
        ],
        topButton: {
            selectors: {
                '.ms-Dialog-button': {
                    color: semanticColors.bodyText,
                },
                '.ms-Dialog-button:hover': {
                    backgroundColor: BaseColors.RED_E00B1C,
                    color: BaseColors.WHITE,
                },
                '.ms-Dialog-button:active': {
                    backgroundColor: BaseColors.RED_E00B1C,
                    color: BaseColors.WHITE,
                },
            },
        },
        innerContent: {
            color: semanticColors.bodyText,
            fontSize: theme.fonts.medium.fontSize,
        },
        subText: {
            color: semanticColors.bodyText,
            fontSize: theme.fonts.medium.fontSize,
        },
    };
};
export var DialogFooterStyles = function (props) {
    return {
        actions: {
            margin: '16px 0 0',
        },
    };
};
