// @fluentui/azure-themes/lib/azure/styles/TagItem.styles.js
import { transparent } from '../Constants';
export var TagItemStyles = function (props) {
    var theme = props.theme, selected = props.selected;
    if (!theme) {
        return {};
    }
    var semanticColors = theme.semanticColors;
    var extendedSemanticColors = semanticColors;
    return {
        root: {
            verticalAlign: 'middle',
            backgroundColor: semanticColors.listItemBackgroundChecked,
            selectors: {
                ':hover': {
                    backgroundColor: semanticColors.listItemBackgroundCheckedHovered,
                },
            },
        },
        text: [
            selected && {
                color: semanticColors.bodyText,
            },
        ],
        close: {
            color: semanticColors.bodyText,
            verticalAlign: 'middle',
            backgroundColor: 'transparent',
            selectors: {
                '&.is-disabled': {
                    backgroundColor: transparent,
                },
                ':active': {
                    backgroundColor: semanticColors.listHeaderBackgroundPressed,
                    color: extendedSemanticColors.controlOutlineHovered,
                },
                ':hover': {
                    backgroundColor: semanticColors.listItemBackgroundCheckedHovered,
                    color: semanticColors.bodyText,
                },
            },
        },
    };
};
