// @fluentui/azure-themes/lib/azure/styles/TagPicker.styles.js
import * as StyleConstants from '../Constants';
export var TagPickerStyles = function (props) {
    var theme = props.theme;
    if (!theme) {
        return {};
    }
    var semanticColors = theme.semanticColors;
    var extendedSemanticColors = semanticColors;
    return {
        root: {
            fontSize: theme.fonts.medium.fontSize,
            border: StyleConstants.borderWidth + " solid " + StyleConstants.transparent,
            backgroundColor: StyleConstants.transparent,
            selectors: {
                '[disabled]': {
                    backgroundColor: semanticColors.buttonBackgroundDisabled,
                    color: semanticColors.buttonTextDisabled,
                },
            },
        },
        input: {
            color: extendedSemanticColors.inputText,
            backgroundColor: StyleConstants.transparent,
            border: 'none',
        },
        itemsWrapper: {
            color: semanticColors.bodyText,
            border: 'none',
            backgroundColor: StyleConstants.transparent,
        },
        text: {
            border: StyleConstants.borderWidth + " solid " + extendedSemanticColors.controlOutline,
        },
    };
};
